### PR TITLE
Kryo serializer

### PR DESCRIPTION
**Pull request by piotr-j**
There is plenty of junk to get rid of, but all tests pass. This is pretty much straight port of json serialiser to kryo. Main problems:

- ```kryo.setReferences(false);``` without it entity references do not work, as all entities are loaded before bags with them
- general write/input stream garbage, not very experienced with that, could use cleanup

Looking forward to feedback. Dont merge as is.

### CODE REVIEW

**Comment by junkdog** [comment]
started taking a look, looking good!

> kryo.setReferences(false);

is this a problem though? this replicates json serialization behavior too.

i noticed the `Bag<String>` bit was commented out. is it because it's serialized as a `Bag<Entity>`?

**Comment by piotr-j** [comment]
Ive fixed the setReferences part, figured out how to disable it only for Entities and it seams to work.
Yes, thats why its commented out. Im not certain that its reasonable to expect only Bag<Enttity> in components. While its same behaviour as json serializers, it would be nice if this stuff worked out of the box more often then not.

**Comment by junkdog** [comment]
Iirc, part of the problem with Bags-of-stuff, was limitations in json - but there might still be a good way to go about it. for the time being - when dealing with serialization, i'm hoping users favor libgdx's collection classes, to avoid conflicts with IntBag/Bag.

**Comment by piotr-j** [comment]
Lets leave the Bags for now, any input on save/load related stuff? Also, ```EntityReferencesTest.java``` how was the json file generated for this? From what Ive read kryo doesnt really work if format changes, so best can be done is roundtrip.

**Comment by junkdog** [comment]
> From what Ive read kryo doesnt really work if format changes, so best can be done is roundtrip.

Ouch, yeah - additive changes to the schema are harmless in json. We might need to address that in the future.

> how was the json file generated for this

hmm, don't recall - I think I just tucked it away so that I could continue testing while serialization was broken. The format with component prefixes was the original format.

**Comment by junkdog** [comment]
```java
protected void save(Writer writer, SaveFileFormat save) {
```
Meh, why did I have to put a Writer in the public interface. Should be OutputStream.

/me returns to testing

**Comment by junkdog** [comment]
fyi, don't merge with master - I introduced a bug...

File: artemis-kryo/src/main/java/com/artemis/io/KryoEntitySerializer.java [hunk 1]
```diff
@@ -230,10 +230,9 @@ protected void clearSerializerCache() {
 		serializers.clear();
 	}
 
-	// TODO do we need to expose this for subclassing?
-	public static class EditFieldSerializer extends FieldSerializer {
+	public static class EditFieldSerializer<T extends Component> extends FieldSerializer<T> {
```
Comment (junkdog) [inline artemis-kryo/src/main/java/com/artemis/io/KryoEntitySerializer.java]: i like this solution; this way, we have a venue for dealing with versioning of component types  + optimized serializers (sometime in the future, we could auto-gen these)

File: artemis-kryo/src/main/java/com/artemis/io/KryoArtemisSerializer.java [hunk 2]
```diff
@@ -66,27 +63,14 @@ public void setKryoInstantiatorStrategy (InstantiatorStrategy strategy) {
 	@Override
 	protected void save(Writer writer, SaveFileFormat save) {
```
Comment (junkdog) [inline artemis-kryo/src/main/java/com/artemis/io/KryoArtemisSerializer.java]: lets just throw an `UnsupportedOperationException` and `@Deprecate` the method. The default case ought to be an Input/OutputStream for load/save.

**Comment by junkdog** [comment]
merging this now; sorry about taking longer than expected.  i ran into some problems with deserialization when dealing with artemis-odb-orion, but all problems turned out to be on my end.

there's probably room to optimize the kryo serialization impl quite a bit (it appears to be ~50% faster than the json backend already, only tested one dataset though). it feels like something which should be directed by proper benchmarks though - i'm mentally preparing to update/refine them. 

:cookie: and thanks ;)

**Comment by piotr-j** [comment]
I had hoped to iron some of the kinds left before merge, but as you wish.

**Comment by junkdog** [comment]
apologies, i felt bad for letting your PR linger. Feel free to continue, and I'll merge in the changes - when you're done.

We should probably have an `artemis-odb-serialization` base module, put all copy-pasted code there. And `@Deprecate` Reader/Writer in favor of Input/OutputStream.

**Comment by piotr-j** [comment]
I suppose main thing now is the gdx dependancy for Base64 stuff used in Reader/Writer. Making a string from raw bytes didnt exactly work when ive tried it. Maybe you have an idea how to do that without encoding first.

**Pull request merged**